wheel/python/clvm_rs/curry_and_treehash.py

```diff
@@ -0,0 +1,145 @@
+        self.null_treehash = shatree_atom(dialect.NULL)
+        self.one_treehash = shatree_atom(ONE)
+
+    # The environment `E = (F . R)` recursively expands out to
+    # `(c . ((q . F) . EXPANSION(R)))` if R is not 0
+    # `1` if R is 0
+
+    def curried_values_tree_hash(self, arguments: List[bytes32]) -> bytes32:
+        if len(arguments) == 0:
+            return self.one_treehash
+
+        inner_curried_values = self.curried_values_tree_hash(arguments[1:])
+
+        return shatree_pair(
+            self.c_kw_treehash,
+            shatree_pair(
+                shatree_pair(self.q_kw_treehash, arguments[0]),
+                shatree_pair(inner_curried_values, self.null_treehash),
+            ),
+        )
+
+    # The curry pattern is `(a . ((q . F)  . (E . 0)))` == `(a (q . F) E)
+    # where `F` is the `mod` and `E` is the curried environment
+
+    def curry_and_treehash(
+        self, hash_of_quoted_mod_hash: bytes32, *hashed_arguments: bytes32
+    ) -> bytes32:
+        """
+        `hash_of_quoted_mod_hash` : tree hash of `(q . MOD)` where `MOD`
+             is template to be curried
+        `arguments` : tree hashes of arguments to be curried
+        """
+
+        for arg in hashed_arguments:
+            if not isinstance(arg, bytes) or len(arg) != 32:
+                raise ValueError(f"arguments must be bytes of len 32: {arg.hex()}")
+
+        curried_values = self.curried_values_tree_hash(list(hashed_arguments))
+        return shatree_pair(
+            self.a_kw_treehash,
+            shatree_pair(
+                hash_of_quoted_mod_hash,
+                shatree_pair(curried_values, self.null_treehash),
+            ),
+        )
+
+    def calculate_hash_of_quoted_mod_hash(self, mod_hash: bytes32) -> bytes32:
+        return shatree_pair(self.q_kw_treehash, mod_hash)
+
+    """
+    Replicates the curry function from clvm_tools, taking advantage of *args
+    being a list.  We iterate through args in reverse building the code to
+    create a clvm list.
+
+    Given arguments to a function addressable by the '1' reference in clvm
+
+    fixed_args = 1
+
+    Each arg is prepended as fixed_args = (c (q . arg) fixed_args)
+
+    The resulting argument list is interpreted with apply (2)
+
+    (2 (1 . self) rest)
+
+    Resulting in a function which places its own arguments after those
+    curried in in the form of a proper list.
+    """
+
+    def curry(self, mod, *args) -> Any:
+        fixed_args: Any = 1
+        for arg in reversed(args):
+            fixed_args = [self.dialect.C_KW, (self.dialect.Q_KW, arg), fixed_args]
+        return [self.dialect.A_KW, (self.dialect.Q_KW, mod), fixed_args]
+
+    """
+    uncurry the given program
+
+    returns `mod, [arg1, arg2, ...]`
+
+    if the program is not a valid curry, return `sexp, NULL`
+
+    This distinguishes it from the case of a valid curry of 0 arguments
+    (which is rather pointless but possible), which returns `sexp, []`
```

Comment:
I wanted to make the API better, and I think this is an improvement because it distinguishes between the two cases of 0 arguments curried (fairly useless but certainly possible) and "not a curried function", and do it without throwing exceptions.

We can create a compatibility layer around this by modifying `Program` in `chia-blockchain` that wraps `uncurry` to have the old functionality (it should be derivable from this). Then we can slowly migrate `import Program` statements to use the `clvm_rs` version of `Program` (and then eventually remove the `chia-blockchain` wrapper).

The main goal is to have the "right" behaviour for the api and eventually using that one everywhere. If you think this is worse than the original, let me know.